File: backend/app/models/resume.py
from __future__ import annotations

import uuid
from typing import Literal

from pydantic import BaseModel, Field


def _new_id() -> str:
    return str(uuid.uuid4())


class RunStyle(BaseModel):
    """Serialized style data for a single text run within a paragraph."""
    text: str
    bold: bool | None = None
    italic: bool | None = None
    underline: bool | None = None
    font_name: str | None = None
    font_size_pt: float | None = None
    color_hex: str | None = None


class Bullet(BaseModel):
    id: str = Field(default_factory=_new_id)
    text: str                          # clean text, bullet char stripped
    run_styles: list[RunStyle]         # per-run formatting for reconstruction
    paragraph_style: str               # original Word paragraph style name
    original_index: int                # position within parent job/section


class FreeParagraph(BaseModel):
    """Non-bullet paragraph within a section (e.g. skill lines, degree lines)."""
    id: str = Field(default_factory=_new_id)
    text: str
    run_styles: list[RunStyle]
    paragraph_style: str
    original_index: int


class Job(BaseModel):
    id: str = Field(default_factory=_new_id)
    title: str
    company: str
    date_range: str
    location: str | None = None
    bullets: list[Bullet] = Field(default_factory=list)
    # Styling for the job header lines (title, company/date)
    title_run_styles: list[RunStyle] = Field(default_factory=list)
    title_paragraph_style: str = "Normal"
    company_run_styles: list[RunStyle] = Field(default_factory=list)
    company_paragraph_style: str = "Normal"
    date_run_styles: list[RunStyle] = Field(default_factory=list)
    date_paragraph_style: str = "Normal"
    # Paragraphs in the job header that didn't resolve into title/company cleanly
    raw_header_paragraphs: list[FreeParagraph] = Field(default_factory=list)
    original_order: int = 0            # position within parent section


SectionType = Literal[
    "contact",
    "summary",
    "experience",
    "education",
    "skills",
    "certifications",
    "projects",
    "military",
    "other",
]


class Section(BaseModel):
    id: str = Field(default_factory=_new_id)
    heading: str
    heading_run_styles: list[RunStyle] = Field(default_factory=list)
    heading_paragraph_style: str = "Normal"
    section_type: SectionType
    jobs: list[Job] = Field(default_factory=list)
    free_paragraphs: list[FreeParagraph] = Field(default_factory=list)
    original_order: int = 0


class DocumentStyles(BaseModel):
    """Document-level style defaults captured from the source DOCX."""
    default_font_name: str | None = None
    default_font_size_pt: float | None = None
    page_width_inches: float | None = None
    page_height_inches: float | None = None
    margin_top_inches: float | None = None
    margin_bottom_inches: float | None = None
    margin_left_inches: float | None = None
    margin_right_inches: float | None = None


ResumeFormat = Literal["chronological", "combination", "unknown"]


class ParsedResume(BaseModel):
    id: str = Field(default_factory=_new_id)
    candidate_name: str = ""
    sections: list[Section]
    detected_format: ResumeFormat = "unknown"
    source_styles: DocumentStyles = Field(default_factory=DocumentStyles)
    parse_warnings: list[str] = Field(default_factory=list)

File: frontend/src/types/resume.ts
// TypeScript mirrors of the backend Pydantic models.
// Keep in sync with backend/app/models/resume.py and analysis.py

export interface RunStyle {
  text: string;
  bold: boolean | null;
  italic: boolean | null;
  underline: boolean | null;
  font_name: string | null;
  font_size_pt: number | null;
  color_hex: string | null;
}

export interface Bullet {
  id: string;
  text: string;
  run_styles: RunStyle[];
  paragraph_style: string;
  original_index: number;
}

export interface FreeParagraph {
  id: string;
  text: string;
  run_styles: RunStyle[];
  paragraph_style: string;
  original_index: number;
}

export interface Job {
  id: string;
  title: string;
  company: string;
  date_range: string;
  location: string | null;
  bullets: Bullet[];
  title_run_styles: RunStyle[];
  title_paragraph_style: string;
  company_run_styles: RunStyle[];
  company_paragraph_style: string;
  date_run_styles: RunStyle[];
  date_paragraph_style: string;
  raw_header_paragraphs: FreeParagraph[];
  original_order: number;
}

export type SectionType =
  | 'contact'
  | 'summary'
  | 'experience'
  | 'education'
  | 'skills'
  | 'certifications'
  | 'projects'
  | 'military'
  | 'other';

export interface Section {
  id: string;
  heading: string;
  heading_run_styles: RunStyle[];
  heading_paragraph_style: string;
  section_type: SectionType;
  jobs: Job[];
  free_paragraphs: FreeParagraph[];
  original_order: number;
}

export interface DocumentStyles {
  default_font_name: string | null;
  default_font_size_pt: number | null;
  page_width_inches: number | null;
  page_height_inches: number | null;
  margin_top_inches: number | null;
  margin_bottom_inches: number | null;
  margin_left_inches: number | null;
  margin_right_inches: number | null;
}

export type ResumeFormat = 'chronological' | 'combination' | 'unknown';

export interface ParsedResume {
  id: string;
  candidate_name: string;
  sections: Section[];
  detected_format: ResumeFormat;
  source_styles: DocumentStyles;
  parse_warnings: string[];
}

// ---- Analysis types (mirrors of analysis.py) ----

export interface BulletScore {
  bullet_id: string;
  relevance_score: number;
  ats_keywords_matched: string[];
  recommendation: 'keep' | 'remove' | 'deprioritize';
  reason: string;
}

export interface JobScore {
  job_id: string;
  relevance_score: number;
  bullet_scores: BulletScore[];
  recommendation: 'keep' | 'collapse' | 'remove';
}

export interface SectionScore {
  section_id: string;
  relevance_score: number;
  job_scores: JobScore[];
  recommendation: 'keep' | 'move_up' | 'move_down' | 'remove';
}

export interface FormatRecommendation {
  recommended_format: 'chronological' | 'combination';
  rationale: string;
  confidence: number;
}

export interface SkillGroup {
  heading: string;
  bullet_ids: string[];
}

export interface AnalysisResult {
  resume_id: string;
  format_recommendation: FormatRecommendation;
  section_scores: SectionScore[];
  ats_keyword_coverage: Record<string, boolean>;
  overall_match_score: number;
  skill_groups: SkillGroup[] | null;
}

// ---- Request / response shapes ----

export interface FormattingSettings {
  format: 'chronological' | 'combination';
  page_limit: 1 | 2;
  font_size_pt: number;
  margins_inches: {
    top: number;
    bottom: number;
    left: number;
    right: number;
  };
}

export interface DownloadRequest {
  resume: ParsedResume;
  selection: Record<string, boolean>;
  section_order: string[];
  job_order: Record<string, string[]>;
  bullet_order: Record<string, string[]>;
  settings: FormattingSettings;
  inline_edits: Record<string, string>;
  skill_groups: SkillGroup[] | null;
}
